node/src/components/contract_runtime.rs

```diff
@@ -259,22 +263,72 @@ impl ContractRuntime {
             + Send,
     {
         match request {
             ContractRuntimeRequest::CommitGenesis {
                 chainspec,
                 responder,
             } => {
-                let result = self.commit_genesis(&chainspec);
-                responder.respond(result).ignore()
+                let engine_state = Arc::clone(&self.engine_state);
+                async move {
+                    let raw_chainspec_bytes = effect_builder.get_chainspec_raw_bytes().await;
+                    let chainspec_registry = {
+                        let mut chainspec_registry = ChainspecRegistry::new();
+                        let chainspec_file_hash =
+                            Digest::hash(raw_chainspec_bytes.chainspec_bytes());
+                        chainspec_registry.insert(CHAINSPEC_RAW.into(), chainspec_file_hash);
+                        if let Some(accounts_bytes) =
+                            raw_chainspec_bytes.maybe_genesis_accounts_bytes()
+                        {
+                            let genesis_accounts_file_hash = Digest::hash(accounts_bytes);
+                            chainspec_registry
+                                .insert(GENESIS_ACCOUNTS_RAW.into(), genesis_accounts_file_hash);
+                        }
+                        chainspec_registry
+                    };
+                    let result =
+                        Self::commit_genesis(&engine_state, &chainspec, chainspec_registry);
+                    responder.respond(result).await
+                }
+                .ignore()
             }
             ContractRuntimeRequest::Upgrade {
-                upgrade_config,
+                mut upgrade_config,
                 responder,
-            } => responder
-                .respond(self.commit_upgrade(*upgrade_config))
-                .ignore(),
+            } => {
+                let engine_state = Arc::clone(&self.engine_state);
+                let metrics = Arc::clone(&self.metrics);
+                async move {
+                    let chainspec_registry = {
+                        let mut chainspec_registry = ChainspecRegistry::new();
+                        let chainspec_raw_bytes = effect_builder.get_chainspec_raw_bytes().await;
+                        let chainspec_file_hash =
+                            Digest::hash(chainspec_raw_bytes.chainspec_bytes());
+                        chainspec_registry.insert(CHAINSPEC_RAW.into(), chainspec_file_hash);
+                        if let Some(global_state_bytes) =
+                            chainspec_raw_bytes.maybe_global_state_bytes()
+                        {
+                            let global_state_toml_hash = Digest::hash(global_state_bytes);
+                            chainspec_registry
+                                .insert(GLOBAL_STATE_RAW.into(), global_state_toml_hash);
+                        }
+                        chainspec_registry
+                    };
```

Comment:
There is some duplication of the code between `ContractRuntimeRequest::Upgrade` and `ContractRuntimeRequest::CommitGenesis`.

If you unify names of bindings (eg. `chainspec_raw_bytes` vs. `raw_chainspec_bytes`) there could be even more duplication :)

Please see if there is a room for some dedup.

Reply:
There was, deduped as suggested